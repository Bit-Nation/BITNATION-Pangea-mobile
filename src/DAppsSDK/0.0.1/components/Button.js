// @flow

import React, { Component } from 'react';
import {
  TouchableOpacity,
  View,
  Text,
} from 'react-native';

import styles from '../../../global/Styles';

type Props = {
  /**
   * @desc Props that should be passed as they are to backed native component.
   */
  nativeProps: Object,
  /**
   * @desc Name of predefined type of component.
   */
  type: string,
  /**
   * @desc Style object.
   */
  style?: Object,
  /**
   * @desc Title of the button.
   */
  title: string,
  /**
   * @desc Style of button title.
   */
  titleStyle?: Object,
  /**
   * @desc Set if button should be disabled.
   */
  disabled: boolean,
  /**
   * @desc Callback to be called when button is pressed.
   */
  onPress: () => any,
}

export default class Button extends Component<Props> {
  static validNativeProps = [
    'disabled',
  ];

  static callbackProps = [
    'onPress',
  ];

  static customProps = [
    'type',
    'title',
    'style',
    'titleStyle',
  ];

  static defaultProps = {
    type: 'transparent',
    disabled: false,
    title: '',
    onPress: () => undefined,
  };

  rootStyleForType(type: string) {
    switch (type) {
      case 'transparent':
        return styles.baseButton;
      case 'action':
        return styles.actionButton;
      case 'custom':
        return null;
      default:
        return undefined;
    }
  }

  titleStyleForType(type: string) {
    switch (type) {
      case 'transparent':
        return styles.buttonTitle;
      case 'action':
        return styles.actionButtonTitle;
      case 'custom':
        return null;
      default:
        return undefined;
    }
  }

  render() {
    const typeStyle = this.rootStyleForType(this.props.type);
    if (typeStyle === undefined) {
      console.warn(`Invalid value '${this.props.type}' for 'type' property of 'Button' component`);
    }

    return (

      <View
        style={[
          typeStyle,
          this.props.style,
        ]}
      >
        <TouchableOpacity
          testID='Touchable'
          {...this.props.nativeProps}
          style={[styles.buttonContainer]}
          onPress={() => this.props.onPress()}
        >
          {
            <Text style={[this.titleStyleForType(this.props.type), this.props.nativeProps.disabled ? styles.disabledButtonTitle : null, this.props.titleStyle]}>
              {this.props.title}
            </Text>
          }
        </TouchableOpacity>
      </View>
    );
  }
}
