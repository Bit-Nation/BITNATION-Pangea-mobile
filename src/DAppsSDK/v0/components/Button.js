// @flow

import React, { Component } from 'react';

import { Button as ReactNativeButton } from 'react-native';
import styles from '../../../global/Styles';

type Props = {
  /**
   * @desc Props that should be passed as they are to backed native component.
   */
  nativeProps: Object,
  /**
   * @desc Name of predefined type of component.
   */
  type: string,
  /**
   * @desc Style object.
   */
  style?: Object
}

const types = [
  'transparent',
  'action',
];


export default class Button extends Component<Props> {
  static validNativeProps = [
    'title',
    'color',
    'disabled',
  ];

  static callbackProps = [
    'onPress',
  ];

  static customProps = [
    'type',
    'style',
  ];

  static defaultProps = {
    type: 'body',
  };

  styleForType(type: string) {
    if (types.includes(type) === false) {
      return null;
    }

    return styles[type];
  }

  render() {
    const typeStyle = this.props.type ? this.styleForType(this.props.type) : null;
    if (this.props.type != null && typeStyle == null) {
      console.warn(`Invalid value '${this.props.type}' for 'type' property of 'Button' component`);
    }

    return (
      <ReactNativeButton
        {...this.props.nativeProps}
        style={[typeStyle, this.props.style]}
      />
    );
  }
}
