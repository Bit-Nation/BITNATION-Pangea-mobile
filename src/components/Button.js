import React, { Component } from 'react';
import { TouchableOpacity, View } from 'react-native';
import { MediaQueryStyleSheet } from 'react-native-responsive';
import PropTypes from 'prop-types';

import Text from './Text';
import Colors from '../global/Colors';

export default class Button extends Component {

  render() {
    const { style, children, onPress, enabled, ...props } = this.props;

    return (
      <View style={[
        styles.baseButton,
        enabled ? styles.enabledButton : styles.disabledButton,
        style]
      } {...props}>
        {
          enabled ?
            <TouchableOpacity style={[styles.container]} onPress={onPress}>
              {children || this._renderTitle()}
            </TouchableOpacity>
            :
            <View style={styles.container}>
              {children || this._renderTitle()}
            </View>
        }

      </View>
    );
  }

  _renderTitle() {
    return (
      <Text buttonTitle={this.props.enabled} disabledButtonTitle={!this.props.enabled}>
        {this.props.title}
      </Text>
    );
  }

}

Button.propTypes = {
  enabled: PropTypes.bool
};

Button.defaultProps = {
  enabled: true,
};

const styles = MediaQueryStyleSheet.create({
  baseButton: {
    borderRadius: 15,
    height: 30,
    justifyContent: 'center',
  },
  enabledButton: {
    backgroundColor: Colors.buttonColor,
  },
  disabledButton: {
    backgroundColor: Colors.disabledButtonColor,
  },
  container: {
    marginLeft: 13,
    marginRight: 13,
  }
});