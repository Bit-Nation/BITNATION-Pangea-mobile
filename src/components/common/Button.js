// @flow

import * as React from 'react';
import {
  TouchableOpacity,
  View,
  Text,
} from 'react-native';
import { MediaQueryStyleSheet } from 'react-native-responsive';

import GlobalStyles from '../../global/Styles';

type Props = {
  /**
   * @desc Style to be applied on top of default styles.
   */
  style?: any,

  styleTitle?: any,
  /**
   * @desc Component(s) to be rendered inside a button.
   */
  children?: React.Node,
  /**
   * @desc Callback to be called on button press.
   */
  onPress?: () => any,
  /**
   * @desc Flag that determines if button is enabled.
   */
  enabled?: boolean,
  /**
   * @desc Title to show on button.
   */
  title?: string,
}

/**
 * @desc Component that renders common button.
 * @return {React.Component} A component.
 */
const Button = ({
  style, children, onPress, enabled, styleTitle, title, ...props
}: Props) => {
  const styles = MediaQueryStyleSheet.create({
    ...GlobalStyles,
  });

  return ((
    <View
      style={[
          styles.baseButton,

          style,
        ]}
      {...props}
    >
      <TouchableOpacity testID='Touchable' disabled={!enabled} style={[styles.buttonContainer]} onPress={onPress}>
        {
        children ||
        <Text style={[styles.buttonTitle, styleTitle, (enabled === false) && styles.disabledButtonTitle]}>
          {title}
        </Text>
      }
      </TouchableOpacity>
    </View>
  ));
};

Button.defaultProps = {
  style: undefined,
  children: undefined,
  enabled: true,
  onPress: () => null,
  title: '',
};

export default Button;
