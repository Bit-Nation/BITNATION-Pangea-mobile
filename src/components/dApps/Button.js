import React, { Component } from 'react';

import { Button as ReactNativeTextButton } from 'react-native';

export default class Button extends Component {
  static validNativeProps = [
    'style',
    'title',
    'color',
  ];

  static stateBasedProps = {
    disabled: 'get',
  };

  static callbackProps = [
    'onPress',
  ];

  render() {
    return (
      <ReactNativeTextButton {...this.props} />
    );
  }
}
