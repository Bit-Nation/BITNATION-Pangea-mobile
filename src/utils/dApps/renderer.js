// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
