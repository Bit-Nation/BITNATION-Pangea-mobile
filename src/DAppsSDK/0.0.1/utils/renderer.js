// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
