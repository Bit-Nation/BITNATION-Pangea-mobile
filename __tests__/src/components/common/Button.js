import React from 'react';
import renderer from 'react-test-renderer';

import Button from '../../../../src/components/common/Button';

describe('Button rendering', () => {

  test('Default button', () => {
    const tree = renderer
      .create(<Button/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Button with title', () => {
    const tree = renderer
      .create(<Button title='Test title'/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Disabled button with title', () => {
    const tree = renderer
      .create(<Button title='Test title' enabled={false}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Enabled button with title', () => {
    const tree = renderer
      .create(<Button title='Test title' enabled={true}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});