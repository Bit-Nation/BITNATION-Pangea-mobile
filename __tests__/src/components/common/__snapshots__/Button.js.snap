// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button rendering Button with title 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#1C497E",
      },
      undefined,
    ]
  }
  title="Test title"
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginLeft": 13,
        "marginRight": 13,
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      buttonTitle={true}
      disabledButtonTitle={false}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
          undefined,
          Object {
            "color": "#72A4DE",
            "fontSize": 14,
            "letterSpacing": -0.02,
            "lineHeight": 19,
            "textAlign": "center",
          },
          false,
          undefined,
        ]
      }
    >
      Test title
    </Text>
  </View>
</View>
`;

exports[`Button rendering Default button 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#1C497E",
      },
      undefined,
    ]
  }
  title=""
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginLeft": 13,
        "marginRight": 13,
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      buttonTitle={true}
      disabledButtonTitle={false}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
          undefined,
          Object {
            "color": "#72A4DE",
            "fontSize": 14,
            "letterSpacing": -0.02,
            "lineHeight": 19,
            "textAlign": "center",
          },
          false,
          undefined,
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Button rendering Disabled button with title 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#37393C",
      },
      undefined,
    ]
  }
  title="Test title"
>
  <View
    style={
      Object {
        "marginLeft": 13,
        "marginRight": 13,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      buttonTitle={false}
      disabledButtonTitle={true}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
          undefined,
          false,
          Object {
            "color": "#5F6D7D",
            "fontSize": 14,
            "letterSpacing": -0.02,
            "lineHeight": 19,
            "textAlign": "center",
          },
          undefined,
        ]
      }
    >
      Test title
    </Text>
  </View>
</View>
`;

exports[`Button rendering Enabled button with title 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#1C497E",
      },
      undefined,
    ]
  }
  title="Test title"
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginLeft": 13,
        "marginRight": 13,
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      buttonTitle={true}
      disabledButtonTitle={false}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
          undefined,
          Object {
            "color": "#72A4DE",
            "fontSize": 14,
            "letterSpacing": -0.02,
            "lineHeight": 19,
            "textAlign": "center",
          },
          false,
          undefined,
        ]
      }
    >
      Test title
    </Text>
  </View>
</View>
`;
